File: lib/ShortestJobFirst.ts
// ShortestJobFirst.ts

/**
 * Define the Process type
 */
export type Process = {
  process_id: number;
  arrival_time: number;
  burst_time: number;
  background: string;
};

export type ScheduledProcess = Process & {
  startTime: number;
  endTime: number;
  waitingTime: number;
  turnaroundTime: number;
};

export type SimulationResult = {
  sequence: Process[]; // For Gantt Chart (includes idle time with process_id -1)
  stats: {
    avgWaitingTime: number;
    avgTurnaroundTime: number;
    cpuUtilization: number;
    throughput: number;
  };
  processStats: ScheduledProcess[];
};

/**
 * Applies the Shortest Job First (SJF) non-preemptive scheduling algorithm.
 */
export function shortestJobFirst(processes: Process[]): SimulationResult {
  const sortedByArrival = [...processes].sort((a, b) => {
    if (a.arrival_time !== b.arrival_time) return a.arrival_time - b.arrival_time;
    return a.process_id - b.process_id;
  });

  const sequence: Process[] = [];
  const processStats: ScheduledProcess[] = [];
  const readyQueue: Process[] = [];
  let currentTime = 0;
  let index = 0;
  const totalBurstTime = processes.reduce((sum, p) => sum + p.burst_time, 0);

  while (index < sortedByArrival.length || readyQueue.length > 0) {
    while (index < sortedByArrival.length && sortedByArrival[index].arrival_time <= currentTime) {
      readyQueue.push(sortedByArrival[index]);
      index++;
    }

    if (readyQueue.length > 0) {
      readyQueue.sort((a, b) => {
        if (a.burst_time !== b.burst_time) return a.burst_time - b.burst_time;
        if (a.arrival_time !== b.arrival_time) return a.arrival_time - b.arrival_time;
        return a.process_id - b.process_id;
      });

      const next = readyQueue.shift()!;
      const startTime = currentTime;
      const waitingTime = startTime - next.arrival_time;
      const turnaroundTime = waitingTime + next.burst_time;
      const endTime = startTime + next.burst_time;

      sequence.push({ ...next, arrival_time: startTime });
      processStats.push({ ...next, startTime, endTime, waitingTime, turnaroundTime });

      currentTime = endTime;
    } else {
      const nextArrival = sortedByArrival[index].arrival_time;
      const idleDuration = nextArrival - currentTime;
      sequence.push({
        process_id: -1,
        arrival_time: -1,
        burst_time: idleDuration,
        background: "transparent",
      });
      currentTime = nextArrival;
    }
  }

  const totalTime = currentTime - (processes.length > 0 ? Math.min(...processes.map(p => p.arrival_time)) : 0);
  const avgWaitingTime = processStats.reduce((sum, p) => sum + p.waitingTime, 0) / (processes.length || 1);
  const avgTurnaroundTime = processStats.reduce((sum, p) => sum + p.turnaroundTime, 0) / (processes.length || 1);
  const cpuUtilization = (totalBurstTime / (totalTime || 1)) * 100;
  const throughput = processes.length / (totalTime || 1);

  return {
    sequence: mergeConsecutive(sequence),
    processStats,
    stats: {
      avgWaitingTime,
      avgTurnaroundTime,
      cpuUtilization,
      throughput
    }
  };
}

/**
 * Merges consecutive blocks of the same process (or idle time) in the sequence.
 */
function mergeConsecutive(sequence: Process[]): Process[] {
  if (sequence.length === 0) return [];

  const merged: Process[] = [{ ...sequence[0] }];

  for (let i = 1; i < sequence.length; i++) {
    const current = sequence[i];
    const last = merged[merged.length - 1];

    if (current.process_id === last.process_id) {
      last.burst_time += current.burst_time;
    } else {
      merged.push({ ...current });
    }
  }

  return merged;
}
  

File: cpp/src/ShortestJobFirst.cpp
#include "Algorithms.h"
#include <algorithm>

std::vector<Process> shortestJobFirst(const std::vector<Process>& processes) {
    // Clone and sort the processes by arrival time
    std::vector<Process> sortedProcesses = processes;
    std::sort(sortedProcesses.begin(), sortedProcesses.end(),
        [](const Process& a, const Process& b) {
            return a.arrival_time < b.arrival_time;
        });

    std::vector<Process> result;
    std::vector<Process> availableProcesses;
    int currentTime = 0;
    size_t index = 0;

    while (index < sortedProcesses.size() || !availableProcesses.empty()) {
        // Move processes that have arrived by currentTime into available processes
        while (index < sortedProcesses.size() &&
               sortedProcesses[index].arrival_time <= currentTime) {
            availableProcesses.push_back(sortedProcesses[index]);
            index++;
        }

        // If there are available processes, pick the one with the shortest burst time
        if (!availableProcesses.empty()) {
            // Sort by burst time to find the shortest job
            std::sort(availableProcesses.begin(), availableProcesses.end(),
                [](const Process& a, const Process& b) {
                    return a.burst_time < b.burst_time;
                });

            Process nextProcess = availableProcesses[0];
            availableProcesses.erase(availableProcesses.begin());

            // Add the selected process to the result
            nextProcess.arrival_time = currentTime;
            result.push_back(nextProcess);

            currentTime += nextProcess.burst_time;
        } else {
            // If no processes are available, create an idle gap
            const Process& nextProcess = sortedProcesses[index];
            int gapDuration = nextProcess.arrival_time - currentTime;
            Process gap(-1, -1, gapDuration, "transparent");
            result.push_back(gap);
            currentTime += gapDuration;
        }
    }

    // Merge consecutive executions of the same process
    std::vector<Process> mergedResult;
    for (size_t i = 0; i < result.size(); i++) {
        const Process& currentProcess = result[i];

        if (!mergedResult.empty() &&
            mergedResult.back().process_id == currentProcess.process_id) {
            mergedResult.back().burst_time += currentProcess.burst_time;
        } else {
            mergedResult.push_back(currentProcess);
        }
    }

    return mergedResult;
}
